framework/source/class/qx/core/MObjectId.js

```diff
@@ -0,0 +1,234 @@
+    owner : {
+      init : null,
+      check : "qx.core.Object",
+      nullable : true,
+      apply : "_applyOwner"
+    },
+
+
+    /** The ID of the object.  */
+    objectId : {
+      init: null,
+      check : "String",
+      nullable : true,
+      apply : "_applyObjectId"
+    }
+  },
+
+  /*
+   * ****************************************************************************
+   * MEMBERS
+   * ****************************************************************************
+   */
+
+  members: {
+    
+    __ownedObjects: null,
+    __changingOwner: false,
+
+    /**
+     * apply owner id
+     */
+    _applyOwner : function(value, oldValue) {
+      if (!this.__changingOwner) {
+        throw new Error("Please use API methods to change owner, not the property");
+      }
+    },
+
+    /**
+     * apply object id
+     */
+    _applyObjectId : function(value, oldValue) {
+      if (this.getOwner() && !this.__changingOwner) {
+        throw new Error("Please use API methods to change owner ID, not the property");
+      }
+      
+      if (typeof this.getContentElement == "function") {
+        var contentElement = this.getContentElement();
+        if (contentElement) {
+          contentElement.updateObjectId();
+        }
+      }
+    },
+    
+    /**
+     * Returns the object with the specified ID
+     * 
+     * @param id
+     *          {String} ID of the object
+     * @return {qx.core.Object?} the found object
+     */
+    getObject: function(id) {
+      if (this.__ownedObjects) {
+        var obj = this.__ownedObjects[id];
+        if (obj !== undefined) {
+          return obj;
+        }
+      }
+      
+      // Handle paths
+      if (id.indexOf('/') > -1) {
+        var segs = id.split('/');
+        var target = this;
+        var found = segs.every(function(seg) {
+          if (!seg.length) {
+            return true;
+          }
+          if (!target) {
+            return false;
+          }
+          var tmp = target.getObject(seg);
+          if (tmp !== undefined) {
+            target = tmp;
+            return true;
+          }
+        });
+        return found ? target : undefined;
+      }
+      
+      // No object, creating the object
+      var obj = this._createObjectImpl(id);
+      if (obj !== undefined) {
+        this.addOwnedObject(obj, id);
+      }
+      
+      return obj;
+    },
+    
+    /**
+     * Creates the object, intended to be overridden. Null is a valid return
+     * value and will be cached by `getObject`, however `undefined` is NOT a
+     * valid value and so will not be cached meaning that `_createObjectImpl`
+     * will be called multiple times until a valid value is returned.
+     * 
+     * @param id {String} ID of the object
+     * @return {qx.core.Object?} the created object
+     */
+    _createObjectImpl: function(id) {
+      return;  // Return undefined
+    },
+    
+    /**
+     * Adds an object as owned by this object
```

Comment:
I don't understand this comment. What does this method actually do?

Reply:
this is adding to heirarchy, eg using the example above, if `this === B` then `this.addOwnedObject(C)` means that `qx.core.Id.getObject("A/B/C") === C`